contracts/defi/provider/MakerProvider.sol

```diff
@@ -0,0 +1,818 @@
+
+    using SafeMath for uint256;
+
+    /* *************** Events *************************** */
+
+    event CdpOpened(address indexed wallet, bytes32 cup, uint256 pethCollateral, uint256 daiDebt);    
+    event CdpUpdated(address indexed wallet, bytes32 cup, uint256 pethCollateral, uint256 daiDebt);    
+    event CdpClosed(address indexed wallet, bytes32 cup);
+
+    /* ********************************** Implementation of Loan ************************************* */
+
+   /**
+     * @dev Opens a collateralized loan.
+     * @param _wallet The target wallet.
+     * @param _collateral The token used as a collateral (must be 0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE).
+     * @param _collateralAmount The amount of collateral token provided.
+     * @param _debtToken The token borrowed (must be the address of the DAI contract).
+     * @param _debtAmount The amount of tokens borrowed.
+     * @param _oracles An array containing the address of MakerDAO's Tub contract as first element.
+     * @return The ID of the created CDP.
+     */
+    function openLoan(
+        BaseWallet _wallet, 
+        address _collateral, 
+        uint256 _collateralAmount, 
+        address _debtToken, 
+        uint256 _debtAmount, 
+        address[] calldata _oracles
+    ) 
+        external 
+        returns (bytes32 _loanId)
+    {
+        require(_collateral == ETH_TOKEN_ADDRESS, "Maker: collateral must be ETH");
+        IMakerCdp makerCdp = IMakerCdp(_oracles[0]);
+        require(_debtToken == makerCdp.sai(), "Maker: debt token must be DAI");
+        _loanId = openCdp(_wallet, _collateralAmount, _debtAmount, makerCdp);
+    }
+
+    /**
+     * @dev Closes a collateralized loan by repaying all debts (plus interest) and redeeming all collateral (plus interest).
+     * @param _wallet The target wallet.
+     * @param _loanId The ID of the target CDP.
+     * @param _oracles An array containing the address of MakerDAO's Tub contract as first element and Uniswap's Factory as second.
+     */
+    function closeLoan(
+        BaseWallet _wallet, 
+        bytes32 _loanId, 
+        address[] calldata _oracles
+    ) 
+        external
+    {
+        closeCdp(_wallet, _loanId, IMakerCdp(_oracles[0]), UniswapFactory(_oracles[1]));
+    }
+
+    /**
+     * @dev Adds collateral to a loan identified by its ID.
+     * @param _wallet The target wallet.
+     * @param _loanId The ID of the target CDP.
+     * @param _collateral The token used as a collateral (must be 0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE).
+     * @param _collateralAmount The amount of collateral to add.
+     * @param _oracles (An array containing the address of MakerDAO's Tub contract as first element.
+     */
+    function addCollateral(
+        BaseWallet _wallet, 
+        bytes32 _loanId, 
+        address _collateral, 
+        uint256 _collateralAmount, 
+        address[] calldata _oracles
+    ) 
+        external
+    {
```

Comment:
Would be nice to add `require(_oracles.length == 1)` as you did for the other providers

Reply:
Agreed. Note that the module always provides the same oracle array for a provider (in the case of Maker it is a length 2), so I've added the condition `require(_oracles.length >= 1` for methods that only use the first entry and `require(_oracles.length == 2` for methods that use both.